### ISSUE TITLE
ccall with a literal struct argument causes ReadOnlyMemoryError()

### ISSUE THREAD

**Issue by peterahrens** [JuliaLang/julia#28338]
As discussed in #28284 and confirmed as a behavior of the latest julia with a new test in #28335, calling ccall with a literal struct argument (made literal by putting the struct in a quotenode) causes a memory error on appveyor `Environment: ARCH=x86_64, JULIA_TEST_MAXRSS_MB=450`. A (failing) test for this case is included in #28335.

The appveyor tests crash with the following message:
```
Error in testset ccall:
Error During Test at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:19
  Got exception LoadError("C:\\projects\\julia\\julia-8f08ff74b0\\share\\julia\\test\\ccall.jl", 572, ReadOnlyMemoryError()) outside of a @test
  LoadError: ReadOnlyMemoryError()
  Stacktrace:
   [1] top-level scope at .\none:0
   [2] eval at .\boot.jl:319 [inlined]
   [3] eval at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:13 [inlined]
   [4] test_struct16(::Type{Main.Test96Main_ccall.Struct16I}, ::Bool) at C:\projects\julia\julia-8f08ff74b0\share\julia\test\ccall.jl:556
   [5] top-level scope at none:0
   [6] include at .\boot.jl:317 [inlined]
   [7] include_relative(::Module, ::String) at .\loading.jl:1052
   [8] include at .\sysimg.jl:29 [inlined]
   [9] include(::String) at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:13
   [10] macro expansion at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:22 [inlined]
   [11] macro expansion at C:\projects\julia\usr\share\julia\stdlib\v0.7\Test\src\Test.jl:1080 [inlined]
   [12] macro expansion at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:21 [inlined]
   [13] macro expansion at .\util.jl:289 [inlined]
   [14] top-level scope at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:19 [inlined]
   [15] top-level scope at .\none:0
   [16] eval at .\boot.jl:319 [inlined]
   [17] #runtests#3(::UInt128, ::Function, ::String, ::String, ::Bool) at C:\projects\julia\julia-8f08ff74b0\share\julia\test\testdefs.jl:25
   [18] #runtests at .\none:0 [inlined] (repeats 2 times)
   [19] (::getfield(Distributed, Symbol("##112#114")){Distributed.CallMsg{:call_fetch}})() at C:\projects\julia\usr\share\julia\stdlib\v0.7\Distributed\src\process_messages.jl:269
   [20] run_work_thunk(::getfield(Distributed, Symbol("##112#114")){Distributed.CallMsg{:call_fetch}}, ::Bool) at C:\projects\julia\usr\share\julia\stdlib\v0.7\Distributed\src\process_messages.jl:56
   [21] macro expansion at C:\projects\julia\usr\share\julia\stdlib\v0.7\Distributed\src\process_messages.jl:269 [inlined]
   [22] (::getfield(Distributed, Symbol("##111#113")){Distributed.CallMsg{:call_fetch},Distributed.MsgHeader,Sockets.TCPSocket})() at .\task.jl:262
  in expression starting at C:\projects\julia\julia-8f08ff74b0\share\julia\test\ccall.jl:572
ERROR: LoadError: Test run finished with errors
in expression starting at C:\projects\julia\julia-8f08ff74b0\share\julia\test\runtests.jl:61
Command exited with code 1
```

**Comment by Keno** [comment]
I cannot reproduce this with the official 1.0 binaries on windows.

**Comment by Keno** [comment]
that still seems to be the case, but I don't see it causing the ReadOnlyMemoryError. For what purpose would the callee be allowed to modify this memory?

**Comment by Keno** [comment]
My conclusion here is that LLVM implements the `byval` attribute incorrectly on win64 (or rather doesn't implement it). In particular, the language reference says:
```
The attribute implies that a hidden copy of the pointee is made between the caller and the callee, so the callee is unable to modify the value in the caller.
```

**Comment by vtjnash** [comment]
> For what purpose would the callee be allowed to modify this memory?

I guess it's callee-owned, so the compiler is permitted to use it as scratch space however it wants?

**Comment by Keno** [comment]
> I guess it's callee-owned, so the compiler is permitted to use it as scratch space however it wants?

Yeah, that was my conclusion also. This is all very ill-described. I'll submit an LLVM patch to fix it, but we can also work around it locally.

**Comment by Keno** [comment]
Candidate patch: https://reviews.llvm.org/D51842

**Comment by Keno** [comment]
Fixes in #29105
